Readme.md

```diff
@@ -0,0 +1,1432 @@
+
+# Table of contents
+
+[TOC]
+
+# General code style (TODO?)
+
+As a general rule, constantly monitor warnings and other notices which your IDE shows you. IntelliJ IDEA in particular has clever analysis features which are very helpful to avoid many mistakes in the code style. If you use IDEA to make commits, enable "Before commit / Perform code analysis" option in its commit window - it will display all errors and warnings it sees in the code you're about to commit.
+
+## Formatting and whitespace
+
+Formatting suggested by the configured IDEA should be preferred. If in doubt, use the "Reformat code" action.
+
+Lines should be indented with two spaces. Lines longer than 120 characters in length should be avoided.
+
+## Naming
+
+### Length of names
+
+Follow the rule "longer names for larger scopes, shorter names for smaller scopes". One-letter names for numeric loops indices are acceptable, same names for most of local variables and pattern variables are not.
+
+### Apostrophes
+
+Avoid using \`s to overload reserved names. Use of apostrophes in pattern matching for equality checking, on the other hand, is acceptable:
+
+```scala
+val name = "hello"
+
+x match {
+  case SomeClass(`name`) => ...
+}
+```
+
+### Constant names
+
+Constants must always follow the `PascalCase` naming rule, according to the default style used in Scala community. Do not use `CAPITAL_SNAKE_CASE` or `camelCase` for constant names.
+
+Constants must always be defined in objects. If a constant is only used in one class, then it should be defined in its companion object. If a constant is used in multiple classes, it should be defined in some common object.
+
+Constants must not be defined in package objects.
+
+### Getters
+
+Do not prefix getters with `get`, just name the getter after the thing it retrieves:
+
+```scala
+trait Nameable {
+  def name: String
+}
+```
+
+### Names shadowing
+
+Avoid clashes of the names with pattern matching, e.g. don't do this:
+
+```scala
+case class User(...)
+case class Organization(owner: User)
+
+val user = User(...)
+...
+organization match {
+  case Organization(user) => ...  // user clashes with the above val
+}
+```
+
+do e.g. this:
+
+```scala
+organization match {
+  case Organization(owner) => ...
+}
+```
+
+Names shadowing can be resolved by the IDEs, but nevertheless is a bad idea because it is very easy to mistake one variable for another. Use more appropriate names for each variable and pattern.
+
+## Imports
+
+### Imports order
+
+Use the IDEA imports optimizer to sort imports. **TODO: add the snippet with the imports configuration** Source code files should not contain unused imports.
+
+### Relative imports
+
+Always use absolute imports; IDEA code style configuration enforces this. Relative imports almost always decrease the readability.
+
+### Imports location
+
+Always put imports at the top of the file instead of nested scopes. In other words, imports must always be grouped together in one place.
```

Comment:
Good question, I should add them explicitly. That's why this section has `TODO?` in it, BTW :)
I think we can use the rules directly from the [Databricks guide](https://github.com/databricks/scala-style-guide#imports). Imports formatting can be enforced easily with the IDEA rules.